_posts/2023-02-28-kani-internship-projects-2022-stubbing.md

```diff
@@ -0,0 +1,361 @@
+This isn't a strange situation since deserialization often leads to bad performance.
+
+```rust
+#[derive(Deserialize)]
+#[cfg_attr(kani, derive(kani::Arbitrary))]
+pub struct MyStruct {
+    // -- snip -- 
+}
+
+#[cfg(kani)]
+fn stub_deserialize<S, T>(_data: &[u8]) -> serde_json::Result<T>
+where
+    T: kani::Arbitrary,
+{
+    Ok(kani::any())
+}
+
+#[cfg(kani)]
+#[kani::proof]
+#[kani::stub(serde_json::from_slice, stub_deserialize)]
+fn verify_with_deserialization() {
+    let data = symbolic_slice();
+    let my_struct = serde_json::from_slice::<MyStruct>(&data);
+    // -- snip -- 
+}
+```
+
+Here, we're using `kani::any()` to generate a symbolic `MyStruct`.
+This is easy to do since we're deriving `kani::Arbitrary` for `MyStruct`.
+
+In principle, this may look like an **over-approximation** because we generate a symbolic value for `MyStruct`.
+But actually, we're missing something:
+[`serde_json::Result<T>`](https://docs.rs/serde_json/latest/serde_json/type.Result.html) can also return a `serde_json::Error` if the deserialization fails.
+However, our `stub_deserialize` function assumes that this case will never happen!
+<!-- I'm not sure this case can be deemed an under-approximation just because of that. -->
+Therefore, the stub we wrote cannot be considered a **sound model** of `serde_json::from_slice`, and it's dangerous to use it in our harnesses.
+
+We hope this convinces you about the risks of stubbing.
+Always keep in mind that stubs are essentially additional assumptions in your harnesses.
+Because of that, we recommend our users to avoid stubbing whenever they can.
+If there's no other option, then you can get more assurance by writing harnesses for your stubs.
+
+## Considered Designs
+
+The experimentation for this feature mainly focused on the program transformation required for stubbing functions, methods and types.
+While the feature had been scoped down to function stubbing, we wanted to keep an eye open for any alternative that allowed us do type stubbing[^footnote-type-stubbing].
+For functions and methods, the transformation boils down to replacing all calls to the original function/method with calls to the replacement function/method.
+
+That said, the transformation can also be applied at different stages of the compilation step.
+As you may know, the Rust language uses multiple [Intermediate Representations](https://rustc-dev-guide.rust-lang.org/overview.html#intermediate-representations) (IRs) to represent and perform analyses on Rust programs.
+In total, we considered five approaches:
+ - Conditional compilation[^footnote-conditional]
+ - Source-to-source transformation
+ - AST-to-AST transformation
+ - HIR-to-HIR transformation
+ - MIR-to-MIR transformation
+
+In summary, we tried to answer the following question for each approach:
+
+> Does the approach allows to perform stubbing of local/external functions/methods/types?
```

Comment:
I guess this is a quote from something, but it's not a well-formed English sentence.  Try
- "Does the approach allow stubbing of ....", or
- "Does the approach allow us to preform stubbing of ..."

Reply:
Done.